nn/transformer.py

```diff
@@ -0,0 +1,231 @@
+"""
+Transformer Modules
+"""
+
+from __future__ import annotations
+from typing import Optional, Any, Union
+from . import relu, layer_norm, dropout, LayerRef, Module, ModuleList, Linear
+
+
+class TransformerEncoderLayer(Module):
+  """
+  Defines one layer of a standard transformer encoder
+  """
+  def __init__(self, dim_model: int, num_heads: int, dim_ff: int = 2048, drop: float = 0.1, activation=relu,
+               layer_norm_eps: float = 1e-5, norm_first: bool = False) -> None:
+    """
+    :param dim_model: hidden dim, PyTorch name: d_model
+    :param num_heads: number heads, PyTorch name: nhead
+    :param dim_ff: dimension of feedforward layer, PyTorch name: dim_feedforward
+    :param drop: Dropout value, PyTorch name: dropout
+    :param activation: activation functional
+    :param layer_norm_eps: Epsilon value for layer normalization
+    :param norm_first: Whether to do layer norm before or afterwards
+    """
+    super().__init__()
+    self.self_attn = MultiheadAttention(dim_model, num_heads, dropout=dropout)  # will change with Attention Modules
+
+    self.linear1 = Linear(dim_ff)
+    self.linear2 = Linear(dim_model)
```

Comment:
Use better names. Maybe like:
```suggestion
    self.linear_ff = Linear(dim_ff)
    self.linear_out = Linear(dim_model)
```